### PR TITLE
[docs] add content to coredb contributor build/test

### PR DESCRIPTION
**Pull request by jaki**

### CODE REVIEW

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  - `sql/yb_foo.sql`: completely new file (for example, with new features)
+  - `sql/yb_pg_foo.sql`: modified version of original PostgreSQL foo.sql (for example, with compatibility edits)
+    - The goal here is to reduce the difference between `foo.sql` and `yb_pg_foo.sql`, when possible.
+
+{{< /tip >}}
+
+### Test flags
+
+The general hierarchy of flags is as follows:
+
```
Comment (ddhodge) [inline docs/content/preview/contribute/core-database/build-and-test.md:209]: capitalize items and add periods

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+1. test framework: the framework may specify some default flags
+1. test superclass: any parent classes of the test class can set flags
+1. test subclass: child classes' flags should take precedence over parent classes'
```
Comment (ddhodge) [inline docs/content/preview/contribute/core-database/build-and-test.md:212]:
```suggestion
1. Test subclass: child class flags should take precedence over parent class flags.
```

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+1. test: a test itself may set flags
+1. user: the user running the test could add flags
+
+There may be some areas where the order of precedence is not followed: help fixing this is welcome.
+
+Here are some ways the user may specify flags to tests (capitals are environment variables):
```
Comment (ddhodge) [inline docs/content/preview/contribute/core-database/build-and-test.md:218]:
```suggestion
Some of the ways in which you can specify flags to tests (capitals are environment variables) are as follows:
```

**Review by ddhodge — APPROVED** [review]
misc suggestions

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+By default, all items are built.
+You can specify a target to narrow down what to build.
+See `./yb_build.sh -h` for the supported targets.
+Besides that, certain flags may skip building some items.
+For example, specifying a flag to run a C++ test skips the Java build (but running a Java test won't skip the C++ build).
+
+Although there is some intelligence to avoid rebuilding parts, it is incomplete.
+For example, the postgres build uses a build stamp calculated from a git commit and working changes; if there is a mismatch, it reruns postgres configure and make.
+Java is worse in that it runs all the build even if there are no changes since the last run.
```
Comment (mbautin) [inline docs/content/preview/contribute/core-database/build-and-test.md:44]: Replace "Java is worse in that it runs all the build" with e.g. "The Java part has a disadvantage that it runs the full build"

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+Use the `--skip-cxx-build`/`--scb` and `--skip-java-build`/`--sj` flags to reduce incremental build times.
+
+`initdb` is a special case because it is only built if it hasn't been built before, but it isn't rebuilt again unless the `reinitdb` target is specified.
+You should run `reinitdb` in case the initial system catalog in the source code is different since the last time `initdb` was run, unless you intentionally want to use the old initial system catalog (for example, to test upgrades).
+
+### Build generator
+
+CMake is used.
+
+If there are any changes to CMake files since the last build, the next incremental build may throw a build error, such as the following:
```
Comment (mbautin) [inline docs/content/preview/contribute/core-database/build-and-test.md:54]: In many cases, CMake knows that some of its input files have changed and re-runs the generation step.

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+```output
+FAILED: build.ninja
```
Comment (mbautin) [inline docs/content/preview/contribute/core-database/build-and-test.md:57]: I have not seen this exact output, and I am not sure what the error is in this case.

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+- fastdebug: fewer compiler optimizations and a few other tweaks on debug
+- asan: address sanitizer
+- tsan: thread sanitizer
+
+### Compiler
+
+By default, `clang` is used, but `gcc` is also supported.
+
+Specify the compiler using the `--gcc<version_number>` and `--clang<version_number>` flags.
+For example, `--gcc11` or `--clang15`.
```
Comment (mbautin) [inline docs/content/preview/contribute/core-database/build-and-test.md:82]: Perhaps don't specify examples with exact versions of GCC and Clang, so we don't have to update these examples.

File: docs/content/preview/contribute/core-database/build-and-test.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+Note that a high parallelism could result in failures if system resources are overused.
+
+### Test frameworks
+
+#### C++ (external) mini cluster
+
+C++ tests are located in `src/yb` or `ent/src/yb` and end in `test.cc`.
+
+Many C++ tests use external mini cluster or mini cluster to create clusters.
+A key difference is that external mini clusters create separate master, tserver, and postgres processes (as in real life), while mini clusters create masters and tservers in memory in the same process and optionally spawn a separate postgres process.
```
Comment (mbautin) [inline docs/content/preview/contribute/core-database/build-and-test.md:167]: Not sure what "in memory" means here. These in-process masters/tservers use disk similarly to a real server process.

File: docs/content/preview/contribute/core-database/build-from-src-centos.md [hunk 9]
```diff
@@ -87,9 +87,9 @@ There is no need to add any of those directories to `PATH`.
 
 ### Build release package
 
-You can build a release package by executing:
+Run the `yb_release` script to build a release package:
 
-```shell
+```output.sh
```
Comment (mbautin) [inline docs/content/preview/contribute/core-database/build-from-src-centos.md:92]: What does this format specifier mean?

**Review by mbautin — APPROVED** [review]

File: docs/content/preview/contribute/core-database/build-from-src-centos.md (hunk 9 above)
Comment (polarweasel) [reply docs/content/preview/contribute/core-database/build-from-src-centos.md:92]: it means "format as shell commands, but don't add a copy button"

Comment (polarweasel) [reply docs/content/preview/contribute/core-database/build-from-src-centos.md:92]: (`nocopy.<language>` does the same thing)

**Pull request merged**